### PR TITLE
[docs] APIs

### PR DESCRIPTION
**Pull request by stevhliu**
This PR expands the API docs to showcase high-level classes and core lower-level building blocks, starting with the optimizers.

- [x] LinearKbit (Linear4bit, Linear8bitLt, etc.)
- [x] Parameters (Params4bit, Int8Params)
- [x] Embeddings (Stable Embedding, Embedding)

### CODE REVIEW

**Comment by stevhliu** [comment]
Hey @Titus-von-Koeller, can you check the docstring descriptions for [AdaGrad](https://moon-ci-docs.huggingface.co/docs/bitsandbytes/pr_1075/en/reference/optim/adagrad) and the base [optimizer](https://moon-ci-docs.huggingface.co/docs/bitsandbytes/pr_1075/en/reference/optim/optim_overview) classes? Since most of the docstring descriptions are the same, if these look good to you, I can easily copy most of them over to the remaining optimizers.

**Comment by Titus-von-Koeller** [comment]
Hey @stevhliu, I'll take a look now. Please be sure to run `pre-commit install` inside your local BNB Git directory.

**Comment by Titus-von-Koeller** [comment]
Everything looks really good, thanks! I think the docstring for the optimizers and base class is good the way you did it.

File: docs/source/reference/optim/adagrad.mdx [hunk 1]
```diff
@@ -1,17 +1,18 @@
 # AdaGrad
 
-[AdaGrad (Adaptive Gradient)](https://jmlr.org/papers/v12/duchi11a.html) is an optimizer that adaptively adjusts the learning rate for each parameter based on their historical gradients.
+[AdaGrad (Adaptive Gradient)](https://jmlr.org/papers/v12/duchi11a.html) is an adaptive learning rate optimizer. AdaGrad stores a sum of the squared past gradients for each parameter and uses it to scale their learning rate. This allows the learning rate to be automatically lower or higher depending on the magnitude of the gradient, eliminating the need to manually tune the learning rate.
```
Comment (Titus-von-Koeller) [inline docs/source/reference/optim/adagrad.mdx:3]: nice summary, much cleaner now!

File: docs/source/reference/optim/adam.mdx [hunk 2]
```diff
@@ -1,15 +1,38 @@
 # Adam
 
+[Adam (Adaptive moment estimation)](https://hf.co/papers/1412.6980) is an adaptive learning rate optimizer, combining ideas from [`SGD`] with momentum and [`RMSprop`] to automatically scale the learning rate:
```
Comment (Titus-von-Koeller) [inline docs/source/reference/optim/adam.mdx:3]: very nice as well, thanks <3

File: docs/source/reference/optim/optim_overview.mdx [hunk 3]
```diff
@@ -1,11 +1,21 @@
 # Overview
 
+[8-bit optimizers](https://hf.co/papers/2110.02861) reduce the memory footprint of 32-bit optimizers without any performance degradation which means you can train large models with many parameters faster. At the core of 8-bit optimizers is block-wise quantization which enables quantization accuracy, computational efficiency, and stability.
```
Comment (Titus-von-Koeller) [inline docs/source/reference/optim/optim_overview.mdx:3]: very nice work, super well done!

File: docs/source/reference/optim/sgd.mdx [hunk 4]
```diff
@@ -1,7 +1,20 @@
 # SGD
 
+Stochastic gradient descent (SGD) is a basic gradient descent optimizer to minimize loss given a set of model parameters and updates the parameters in the opposite direction of the gradient. The update is performed on a randomly sampled mini-batch of data from the dataset.
+
+bitsandbytes also supports momentum and Nesterov momentum to accelerate SGD by adding a weighted average of past gradients to the current gradient.
```
Comment (Titus-von-Koeller) [inline docs/source/reference/optim/sgd.mdx:5]: This docs effort is so useful. It's these little details that makes the users more likely to find and use functionalities, really cool!

File: docs/source/reference/nn/embeddings.mdx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Embedding
+
+The embedding class is used to store and retrieve word embeddings from their indices. There are two types of embeddings in bitsandbytes, the standard PyTorch [`Embedding`] class and the [`StableEmbedding`] class.
```
Comment (Titus-von-Koeller) [inline docs/source/reference/nn/embeddings.mdx:3]: clear and to the point this section!

**Pull request merged**